include/pfn/expected.hpp

```diff
@@ -0,0 +1,1483 @@
+
+public:
+  constexpr unexpected(unexpected const &) = default;
+  constexpr unexpected(unexpected &&) = default;
+
+  template <class Err = E>
+  constexpr explicit unexpected(Err &&e) noexcept(::std::is_nothrow_constructible_v<E, Err>)
+    requires(not ::std::is_same_v<::std::remove_cvref_t<Err>, unexpected> &&        //
+             not ::std::is_same_v<::std::remove_cvref_t<Err>, ::std::in_place_t> && //
+             ::std::is_constructible_v<E, Err>)
+      : e_(FWD(e))
+  {
+  }
+
+  template <class... Args>
+  constexpr explicit unexpected(::std::in_place_t, Args &&...a) noexcept(::std::is_nothrow_constructible_v<E, Args...>)
+    requires ::std::is_constructible_v<E, Args...>
+      : e_(FWD(a)...)
+  {
+  }
+
+  template <class U, class... Args>
+  constexpr explicit unexpected(::std::in_place_t, ::std::initializer_list<U> i, Args &&...a) noexcept(
+      ::std::is_nothrow_constructible_v<E, ::std::initializer_list<U> &, Args...>)
+    requires ::std::is_constructible_v<E, ::std::initializer_list<U> &, Args...>
+      : e_(i, FWD(a)...)
+  {
+  }
+
+  constexpr unexpected &operator=(unexpected const &) = default;
+  constexpr unexpected &operator=(unexpected &&) = default;
+
+  constexpr E const &error() const & noexcept { return e_; };
+  constexpr E &error() & noexcept { return e_; };
+  constexpr E const &&error() const && noexcept { return ::std::move(e_); };
+  constexpr E &&error() && noexcept { return ::std::move(e_); };
+
+  constexpr void swap(unexpected &other) noexcept(::std::is_nothrow_swappable_v<E>)
+  {
+    static_assert(::std::is_swappable_v<E>);
+    using ::std::swap;
+    swap(e_, other.e_);
+  }
+
+  template <class E2> constexpr friend bool operator==(unexpected const &x, unexpected<E2> const &y)
+  {
+    return x.e_ == y.e_;
+  }
+
+  constexpr friend void swap(unexpected &x, unexpected &y) noexcept(noexcept(x.swap(y)))
+    requires ::std::is_swappable_v<E>
+  {
+    x.swap(y);
+  }
+
+private:
+  E e_;
+};
+
+template <class E> unexpected(E) -> unexpected<E>;
+
+template <class T, class E> class expected;
+namespace detail {
+template <typename> constexpr bool _is_some_expected = false;
+template <typename T, typename E> constexpr bool _is_some_expected<::pfn::expected<T, E>> = true;
+
+template <typename T, typename E>
+constexpr bool _is_valid_expected =                                   //
+    not ::std::is_reference_v<T>                                      //
+    && not ::std::is_function_v<T>                                    //
+    && not ::std::is_same_v<::std::remove_cv_t<T>, ::std::in_place_t> //
+    && not ::std::is_same_v<::std::remove_cv_t<T>, unexpect_t>        //
+    && not _is_some_unexpected<::std::remove_cv_t<T>>                 //
+    && detail::_is_valid_unexpected<E>;
+} // namespace detail
+
+// declare void specialization
+template <class T, class E>
+  requires ::std::is_void_v<T>
+class expected<T, E>;
+
+template <class T, class E> class expected {
```

Comment:
Shouldn't there be `has_val`, `val`, and `unex` members?
https://en.cppreference.com/w/cpp/utility/expected.html

Reply:
No; these names are "exposition only", see bottom of https://eel.is/c++draft/expected#object.general . They are meant to provide names for entities used in the standard wording, but since they are private, the actual names used inside the implementation can be (and typically are) different.